### PR TITLE
MAINT: Tidy macros in scalar_new

### PR DESCRIPTION
**Pull request by eric-wieser**
Rather than trying to keep track of weird tri-state `@default@` and `@work@` variables, this uses a primitive dictionary with types as keys.

---

@seberg, is this the type of cleanup you had in mind in https://github.com/numpy/numpy/pull/15393#pullrequestreview-347012965

### CODE REVIEW

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 1]
```diff
@@ -2555,34 +2556,36 @@ object_arrtype_dealloc(PyObject *v)
...
- * #default = 0*17,1*2,2#
  */
 
-#define _NPY_UNUSED2_1
-#define _NPY_UNUSED2_z
-#define _NPY_UNUSED2_0 NPY_UNUSED
-#define _NPY_UNUSED1_0
-#define _NPY_UNUSED1_1
-#define _NPY_UNUSED1_2 NPY_UNUSED
+#if @TYPE@_IS_OBJECT
```
Comment (seberg) [inline numpy/core/src/multiarray/scalartypes.c.src]: I suppose these should be `#if _@TYPE@_IS_OBJECT`? (Was going to add `defined` first, but that is unnecessary I guess).
I think I like this style, although did not look through all yet. IIRC one thing that I was wondering if we cannot hardcode the result of `PyTuple_GET_ITEM(cls.tp_bases, num)`

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: Good catch, thanks

**Review by eric-wieser — COMMENTED** [review]
Updated to use `defined()`, I think it's a little more understandable that way.

**Comment by seberg** [comment]
Thanks, the dance around `NPY_UNUSED` is unfortunate, but I prefer this style, so putting it in.

**Pull request merged**